### ISSUE TITLE
RPI 5 gpio numbers

### ISSUE THREAD

**Issue by livep2000** [fivdi/onoff#199]
Recently I burned out my rpi4, just at the moment my highly complicated radio transceiver reached some point of functionality. 
Ordered a pi5, to discover its not working. After reading and reading this issues tracker give the fairly simple solution.

Just add 571 to the gpio pin number. Proposing something like this to make it fully compatible:

```
let file = fs.readFileSync("/proc/cpuinfo", "utf8");
let arr = file.split(/\r?\n/);
arr.forEach((line, idx)=> {
    if(line.includes("Pi 5")){
        console.log(line);
        gpio += 571;
        }
});
```
Imre

**Comment by SamDecrock** [comment]
I don't think it's Pi5 related but rather the version of Raspbian. On Bullseye it works, on Bookworm it's broken.

**Comment by nwpray** [comment]
I opened a PR to lookup the base address and add that automatically to the address you pass
https://github.com/fivdi/onoff/pull/204

**Comment by gdorbes** [comment]
The gpio number issue is a matter of both RPi model and os version. My tests:

- With Bookworm on RPI5B, offset is 571.
- With Bookworm on RPi4B, offset is 522.

To find the right value, see `cat /sys/kernel/debug/gpio`

**Comment by SamDecrock** [comment]
In essence, there is nothing wrong with the `onoff` library.

It's just that the gpio numbers are different, so instead of 1, 2, 3, 4, ... it's 571, 572, 573, 574, ...

All you have to do is make sure you use the new numbering, eg:

```js
const gpio0 = new Gpio(571, 'out', { activeLow: true });
```

**Comment by gdorbes** [comment]
See https://www.npmjs.com/package/rpi-io a new module that supports recent versions of Raspberry and related OS and middleware.

**Comment by fq-selbach** [comment]
> In essence, there is nothing wrong with the `onoff` library.
> 

Well, the **onoff** library relies on the sysfs interface (/sys/class/gpio). This interface was officially deprecated years ago, but in Trixie (and for all Raspberry Pi 5 hardware), it has been removed or heavily restricted in favor of **libgpiod**.
Maybe you can make it work on current systems, but its a sinking ship :-(

**Comment by gdorbes** [comment]
> > In essence, there is nothing wrong with the `onoff` library.
> 
> Well, the **onoff** library relies on the sysfs interface (/sys/class/gpio). This interface was officially deprecated years ago, but in Trixie (and for all Raspberry Pi 5 hardware), it has been removed or heavily restricted in favor of **libgpiod**. Maybe you can make it work on current systems, but its a sinking ship :-(

I agree with 95% of your comment. It is the reason why I have developped  a library based on libgpiod for recent RPi hardware and OS. See https://github.com/gdorbes/rpi-io 😄 

For the remaining 5%, I must admit that I still use sysf interface to manage hardware-based PWM interface. I did not find a "modern" solution to replace sysf. Of course software-based PWM can be used but it is less effective for high frequency pulse. Any idea is welcome. Thanks

**Comment by fq-selbach** [comment]
> I agree with 95% of your comment. It is the reason why I have developped a library based on libgpiod for recent RPi hardware and OS. See https://github.com/gdorbes/rpi-io 😄

I'm looking at it right now and will try to use it as an 'onoff' replacement :-)

> For the remaining 5%, I must admit that I still use sysf interface to manage hardware-based PWM interface. I did not find a "modern" solution to replace sysf.

As far as I understand it the PWM part of sysfs (`/sys/class/pwm/`) is actually NOT deprecated, only the GPIO part (`/sys/class/gpio`). Its a bit confusing, so I'm not 100% certain -_-

**Comment by SamDecrock** [comment]
> > In essence, there is nothing wrong with the `onoff` library.
> 
> Well, the **onoff** library relies on the sysfs interface (/sys/class/gpio). This interface was officially deprecated years ago, but in Trixie (and for all Raspberry Pi 5 hardware), it has been removed or heavily restricted in favor of **libgpiod**. Maybe you can make it work on current systems, but its a sinking ship :-(

Well, that's a different issue. 

This issue is about the gpio numbers being different on Bookworm vs Bullseye.

Your issue is that `/sys/class/gpio` is not available anymore in Trixie ;-) I created a new ticket for that https://github.com/fivdi/onoff/issues/206

**Comment by OlliMartin** [comment]
As far as I can tell this is not related to the library because you can change the offset, but for anyone stumbling over this as well: 
I cleanly installed a raspberry pi5 16gb with the official raspberry image 2-3 days back - offset seems to (sometimes?) be 569, not 571.

```
$ cat /sys/kernel/debug/gpio

// [..]
gpiochip0: GPIOs 569-622, parent: platform/1f000d0000.gpio, pinctrl-rp1:
 gpio-569 (ID_SDA              )
 gpio-570 (ID_SCL              )
```
